ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -380,43 +350,66 @@ static void RefreshRootVnodeIfNecessary_Locked(VirtualizationRootHandle rootHand
     rootEntry.rootVNodeVid = vid;
 }
 
-// Returns negative value if it failed, or inserted index on success
-KEXT_STATIC VirtualizationRootHandle InsertVirtualizationRoot_Locked(PrjFSProviderUserClient* userClient, pid_t clientPID, vnode_t vnode, uint32_t vid, FsidInode persistentIds, const char* path)
+KEXT_STATIC VirtualizationRootHandle FindOrInsertVirtualizationRoot_LockedMayUnlock(vnode_t virtualizationRootVNode, uint32_t rootVid, FsidInode persistentIds, const char* path)
 {
-    VirtualizationRootHandle rootIndex = FindUnusedIndexOrGrow_Locked();
-    
-    if (RootHandle_None != rootIndex)
+    VirtualizationRootHandle rootIndex;
+    do
     {
-        assert(rootIndex < s_maxVirtualizationRoots);
-        assert(!s_virtualizationRoots[rootIndex].inUse);
-        
-        // Retain a strong reference to the vnode if we have an active provider on it
-        if (userClient != nullptr)
+        rootIndex = FindRootAtVnode_Locked(virtualizationRootVNode, rootVid, persistentIds);
+        if (rootIndex >= 0)
         {
-            errno_t error = vnode_get(vnode);
-            assertf(error == 0, "The calling code should already hold an iocount on vnode, so vnode_get should never fail. error = %d", error);
+            return rootIndex;
         }
         
-        VirtualizationRoot* root = &s_virtualizationRoots[rootIndex];
-        
-        root->providerUserClient = userClient;
-        root->providerPid = clientPID;
-        root->inUse = true;
-
-        root->rootVNode = vnode;
-        root->rootVNodeVid = vid;
-        KextLog_File(vnode, "InsertVirtualizationRoot_Locked: virtualization root inserted at index %d: (path: \"%s\", fsid: 0x%x:%x, inode: 0x%llx) directory vnode %p:%u, user client PID %d, IOUC %p.",
-            rootIndex, path, persistentIds.fsid.val[0], persistentIds.fsid.val[1], persistentIds.inode, KextLog_Unslide(vnode), vid, clientPID, KextLog_Unslide(userClient));
-        
-        root->rootFsid = persistentIds.fsid;
-        root->rootInode = persistentIds.inode;
-
-        if (path != nullptr)
+        rootIndex = FindUnusedIndex_Locked();
+        if (rootIndex < 0)
         {
-            strlcpy(root->path, path, sizeof(root->path));
+            // No space, resize array
+            uint16_t newLength = MIN(s_maxVirtualizationRoots * 2u, INT16_MAX + 1u);
+            if (newLength <= s_maxVirtualizationRoots)
+            {
+                KextLog_Error("FindOrInsertVirtualizationRoot_LockedMayUnlock: growing virtualization root array for root at '%s' failed: out of valid indices, array has reached %u items\n", path, s_maxVirtualizationRoots);
+                return RootHandle_None;
+            }
+        
+            // Must drop lock to safely allocate memory with blocking alloc
+            RWLock_ReleaseExclusive(s_virtualizationRootsLock);
+            VirtualizationRoot* grownArray = Memory_AllocArray<VirtualizationRoot>(newLength);
```

Comment:
Check if `Memory_AllocArray` returned null?

Reply:
For blocking allocs, this should pretty much never happen, and the caller currently asserts on this function succeeding, but I guess it's good to be explicit about it. I've added that check now, it'll be in the next push.